pravega/pravega
# Issue 2262: (SegmentStore) Admin tools for SegmentContainer Recovery

**NOTE: this includes PRs #2228  and #2259. Those should be reviewed and merged before looking at this PR.**

**Change log description**
Implemented Admin Tools for the following:
- Enabling a `BookKeeperLog` that is currently disabled.
- Disabling a `BookKeeperLog` that is currently enabled.
- Executing a non-invasive `DurableLog` recovery for a specified container
    - Non-invasive means it is performed without the need to fence out Tier1 or otherwise make any sort of modifications to the data in Tier1.
    - `DurableLog` recovery means we are only reading data from Tier1, deserializing `Operations` and adding them a `ReadIndex` (no cache) and validating Metadata. This is the exact same process as in a normal SegmentContainer recovery, but we don't start other sub-services, such as `OperationProcessor` or `StorageWriter`.

**Purpose of the change**
Fixes #2262.

**What the code does**
- New commands have been added to the `admin` package of the Segment Store.
- Recovery code inside `DurableLog` has been factored out
- All admin tool access is done via "Debug Wrappers" and no changes to actual functionality has been made to the greatest extent possible.

**How to verify it**
Similar to #2256 (so this will be more succinct):
1. In a separate terminal, type `./gradlew test:integration:selftest -Dtarget=OutOfProcess -Dreads=false -Dc=4 -Ds=1 -Dsc=16 -Dws=1000 -Do=1000000 -Dp=100 -Dpp=100 -Dpause=true`
    - Wait until you are prompted to press any key to continue
2. Create a new config file to use with the admin tool or make sure these config values are passed to the tool
```
pravegaservice.containerCount=4
pravegaservice.zkURL=localhost:9000
bookkeeper.bkLedgerPath=/pravega/selftest/bookkeeper/ledgers
bookkeeper.zkMetadataPath=/pravega/selftest/segmentstore/containers
bookkeeper.bkEnsembleSize=1
bookkeeper.bkAckQuorumSize=1
bookkeeper.bkWriteQuorumSize=1
```
3. Run the Admin tools: 
    -`./gradlew segmentstore:server:host:admin` - you'll need to enter the config props manually.
    -`./gradlew segmentstore:server:host:admin -Dpravega.configurationFile=config.properties`
4. Type the following commands
    - `bk list` - see all available logs
    - `bk disable 0` - disable log 0 (NOTE: this will cause your running Pravega instance to shutdown container 0 and restart it in offline mode)
    - `bk enable 0` - re-enable log 0 (in a couple of minutes, your Running Pravega instance will resume its normal operations
    - `container recover 0` - attempt local non-intrusive recovery of Container 0. It really doesn't matter if the log is enabled or not; if it's disabled you are guaranteed to get all the latest data, othewise this might not display the latest one (that's how BK works)
